###  Grouping Data & Aggregating Variables

In [25]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [26]:
#Creating string of project folder.

path = r'/Users/nautis1/data_project/09-2025-Instacart-Basket-Analysis/Instacart Basket Analysis'

In [27]:
#importing orders_products_combined dataframe from the pickle file
orders_products_combined = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'ords_prods_merge_new_var.pkl'))


In [28]:
#cleaning prices column 
orders_products_combined.loc[orders_products_combined['prices'] >100, 'prices'] = np.nan


#### Task 2

In [29]:
#find aggregated mean of order_number grouped by department_id
orders_products_combined.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   3139998      138    prior            28          6                 11   
1   1977647      138    prior            30          6                 17   
2    389851      709    prior             2          0                 21   
3    652770      764    prior             1          3                 13   
4   1813452      764    prior             3          4                 17   

   days_since_prior_order  customer_first_order  add_to_cart_order  reordered  \
0                     3.0                 False                  5          0   
1                    20.0                 False                  1          1   
2                     6.0                 False                 20          0   
3                     NaN                  True                 10          0   
4                     9.0                 False                 11          1   

  _merge        price_label     busiest_day    busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy  Regularly busy   
2   both  Mid-range product     Busiest day    Busiest days   
3   both  Mid-range product  Regularly busy      Least busy   
4   both  Mid-range product      Least busy      Least busy   

  busiest_period_of_day  
0           Most orders  
1           Most orders  
2        Average orders  
3           Most orders  
4           Most orders

In [30]:
orders_products_combined.groupby('department_id').agg({ 'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### Task 3

Overall Trends: The entire dataframe generally has higher mean order numbers across departments, indicating it captures a broader range of customer behaviours than the subset.
Notable Differences: Department 21 has the highest mean in both, but it’s higher in the subset (25.53) than the entire dataset (22.90). Department 10 greatly increases the dataframe as a whole (20.19 vs. 18.68).
Consistent Departments: Some departments, like 13 and 17, show little difference between the two datasets, while others, like 3, 9, and 12, have notable variations.
Overall, the entire dataframe reveals more frequent ordering patterns across departments compared to the subset.

#### Task 4

In [31]:
# create max_order column
orders_products_combined['max_order'] = orders_products_combined.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/m1/lrlcgb215jd9lw8n87qhqq_w0000gn/T/ipykernel_94333/2447589412.py:2: FutureWarning: The provided callable <function max at 0x112ef6160> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  orders_products_combined['max_order'] = orders_products_combined.groupby(['user_id'])['order_number'].transform(np.max)


In [32]:
orders_products_combined.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  ...  \
0   3139998      138    prior            28          6  ...   
1   1977647      138    prior            30          6  ...   
2    389851      709    prior             2          0  ...   
3    652770      764    prior             1          3  ...   
4   1813452      764    prior             3          4  ...   

   days_since_prior_order  customer_first_order  add_to_cart_order  reordered  \
0                     3.0                 False                  5          0   
1                    20.0                 False                  1          1   
2                     6.0                 False                 20          0   
3                     NaN                  True                 10          0   
4                     9.0                 False                 11          1   

   _merge        price_label     busiest_day    busiest_days  \
0    both  Mid-range product  Regularly busy  Regularly busy   
1    both  Mid-range product  Regularly busy  Regularly busy   
2    both  Mid-range product     Busiest day    Busiest days   
3    both  Mid-range product  Regularly busy      Least busy   
4    both  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order  
0           Most orders        32  
1           Most orders        32  
2        Average orders         5  
3           Most orders         3  
4           Most orders         3  

[5 rows x 21 columns]

In [33]:
#create loyalty_flag
orders_products_combined.loc[orders_products_combined['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [34]:
#create loyalty_flag
orders_products_combined.loc[(orders_products_combined['max_order'] <= 40) & (orders_products_combined['max_order'] > 10) , 'loyalty_flag'] = 'Regular customer'

In [35]:
#create loyalty_flag
orders_products_combined.loc[orders_products_combined['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [36]:
orders_products_combined['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

#### Task 5

In [37]:
# get descriptive statistics for customer spending across categories
orders_products_combined.groupby(['loyalty_flag']).agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773575  1.0  25.0
New customer      7.801206  1.0  25.0
Regular customer  7.798262  1.0  25.0

The prices statistics reveals opposite pattern than number of orders meaning loyal customers have most number of orders but lowest mean which might be due to the number of orders they place the mean gets on the lower side where as for regular customers since the number of orders are less an outlier on the higher side of price will spike the mean considerably. 

#### Task 6

In [38]:
#find the avg price across all the orders by the user
orders_products_combined['avg_price'] = orders_products_combined.groupby(['user_id'])['prices'].transform(np.average)

In [39]:
orders_products_combined.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  ...  \
0   3139998      138    prior            28          6  ...   
1   1977647      138    prior            30          6  ...   
2    389851      709    prior             2          0  ...   
3    652770      764    prior             1          3  ...   
4   1813452      764    prior             3          4  ...   

   add_to_cart_order  reordered  _merge        price_label     busiest_day  \
0                  5          0    both  Mid-range product  Regularly busy   
1                  1          1    both  Mid-range product  Regularly busy   
2                 20          0    both  Mid-range product     Busiest day   
3                 10          0    both  Mid-range product  Regularly busy   
4                 11          1    both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order      loyalty_flag avg_price  
0  Regularly busy           Most orders        32  Regular customer  6.935811  
1  Regularly busy           Most orders        32  Regular customer  6.935811  
2    Busiest days        Average orders         5      New customer  7.930208  
3      Least busy           Most orders         3      New customer  4.972414  
4      Least busy           Most orders         3      New customer  4.972414  

[5 rows x 23 columns]

In [40]:
#create a spending flag for low spender
orders_products_combined.loc[orders_products_combined['avg_price'] < 10, 'spending_flag'] = 'Low Spender'

In [41]:
#populate spending flag for high spender
orders_products_combined.loc[orders_products_combined['avg_price'] >= 10, 'spending_flag'] = 'High Spender'

In [42]:
orders_products_combined['spending_flag'].value_counts(dropna= False)

spending_flag
Low Spender     31770649
NaN               515228
High Spender      118982
Name: count, dtype: int64

#### Task 7

In [43]:
#find the mediam days_since_prior order for all users
orders_products_combined['median_days'] = orders_products_combined.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/m1/lrlcgb215jd9lw8n87qhqq_w0000gn/T/ipykernel_94333/3438866895.py:2: FutureWarning: The provided callable <function median at 0x1134d3e20> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  orders_products_combined['median_days'] = orders_products_combined.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [44]:
#create regularity flag 
orders_products_combined.loc[orders_products_combined['median_days'] > 20, 'regularity_flag'] = 'Non-frequent customer'

In [45]:
#create regularity flag for regular customer
orders_products_combined.loc[(orders_products_combined['median_days'] > 10) & (orders_products_combined['median_days'] <=20), 'regularity_flag'] = 'Regular customer'

In [46]:
#create flag for Frequent customer
orders_products_combined.loc[orders_products_combined['median_days'] <= 10, 'regularity_flag'] = 'Frequent Customer'

In [47]:
orders_products_combined['regularity_flag'].value_counts(dropna= False)

regularity_flag
Frequent Customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

#### Task 9

In [48]:
# Exporting the Dataframe
orders_products_combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))